### PR TITLE
Add closeVault dftx

### PR DESCRIPTION
**Pull request by jingyi2811**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Add closeVault dftx

#### Which issue(s) does this PR fixes?:

Fixes part of #532

### CODE REVIEW

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_governance_createcfp.test.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ import { GovernanceMasterNodeRegTestContainer } from '../../../jellyfish-api-cor
 import { governance } from '@defichain/jellyfish-api-core'
 import { RegTest } from '@defichain/jellyfish-network'
 
-describe('createCfp', () => {
+describe.skip('createCfp', () => {
```
Comment (canonbrother) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_governance_createcfp.test.ts:13]: hmm... what happen on the `skip` huh??

**Review by canonbrother — COMMENTED** [review]
kindly tests on closevault on state `mayLiquidate` and `frozen`

File: packages/jellyfish-transaction/src/script/dftx/dftx.ts [hunk 2]
```diff
@@ -214,8 +216,8 @@ export class CDfTx extends ComposableBuffer<DfTx<any>> {
         return compose<ICXCloseOrder>(CICXCloseOrder.OP_NAME, d => new CICXCloseOrder(d))
       case CICXCloseOffer.OP_CODE:
         return compose<ICXCloseOffer>(CICXCloseOffer.OP_NAME, d => new CICXCloseOffer(d))
-      case CCreateCfp.OP_CODE:
-        return compose<CreateProposal>(CCreateCfp.OP_NAME, d => new CCreateCfp(d))
+      // case CCreateCfp.OP_CODE:
```
[the remark this answers is not in the document]
Comment (canonbrother) [reply packages/jellyfish-transaction/src/script/dftx/dftx.ts:219]: its fixed now

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_loans/CloseVault.test.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    const stack = toOPCodes(
+      SmartBuffer.fromBuffer(Buffer.from(hex, 'hex'))
+    )
+    const buffer = toBuffer(stack)
+    expect(buffer.toString('hex')).toStrictEqual(hex)
+    expect((stack[1] as OP_DEFI_TX).tx.type).toStrictEqual(0x65)
+  })
+})
+
+const header = '6a3c4466547865' // OP_RETURN(0x6a) (length 3c = 0x60) CDfTx.SIGNATURE(0x44665478) CCloseVault.OP_CODE(0x65)
```
Comment (ivan-zynesis) [inline packages/jellyfish-transaction/__tests__/script/dftx/dftx_loans/CloseVault.test.ts:45]: Minor comment bug. 0x3c = 60 decimal

**Review by ivan-zynesis — COMMENTED** [review]
Looks good, will revisit when test unskipped and extra test case requested by @canonbrother added

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_loans/CloseVault.test.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+  CloseVault
+} from '../../../../src/script/dftx/dftx_loans'
+import { OP_CODES } from '../../../../src/script'
+import { toBuffer, toOPCodes } from '../../../../src/script/_buffer'
+import { OP_DEFI_TX } from '../../../../src/script/dftx'
+
+it('should bi-directional buffer-object-buffer', () => {
+  const fixtures = [
+    /**
+     * loan : {
```
Comment (surangap) [inline packages/jellyfish-transaction/__tests__/script/dftx/dftx_loans/CloseVault.test.ts:13]:
```suggestion
     * closeVault : {
```

File: packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_close_vault.test.ts [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+  })
+
+  it('should not closeVault by anyone other than the vault owner', async () => {
+    const script = await providers.elliptic.script()
+    const fromScript = P2WPKH.fromAddress(RegTest, await tGroup.get(1).generateAddress(), P2WPKH).getScript()
+
+    const txn = await builder.loans.closeVault({
+      vaultId: vaultToBeClosedId,
+      to: fromScript
+    }, script)
```
Comment (surangap) [inline packages/jellyfish-transaction-builder/__tests__/txn/txn_builder_loan_close_vault.test.ts:329]:
```suggestion
    const toScript = P2WPKH.fromAddress(RegTest, await tGroup.get(1).generateAddress(), P2WPKH).getScript()

    const txn = await builder.loans.closeVault({
      vaultId: vaultToBeClosedId,
      to: toScript
    }, script)
```

File: packages/testcontainers/src/containers/RegTestContainer/Masternode.ts [hunk 6]
```diff
@@ -210,6 +224,48 @@ export class MasterNodeRegTestContainer extends RegTestContainer {
...
 
+  /**
+   * Wait for active price
+   *
+   * @param {string} fixedIntervalPriceId
+   * @param {number} activePrice
+   * @param {number} [timeout=30000] in ms
+   * @return {Promise<void>}
+   */
+  async waitForActivePrice (fixedIntervalPriceId: string, activePrice: number, timeout = 30000): Promise<void> {
```
Comment (surangap) [inline packages/testcontainers/src/containers/RegTestContainer/Masternode.ts:235]: duplicate of `waitForNextPrice` ?🤔

Comment (jingyi2811) [reply packages/testcontainers/src/containers/RegTestContainer/Masternode.ts:235]: Next price and active price are different

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts [hunk 7]
```diff
@@ -173,4 +174,19 @@ export class TxnBuilderLoans extends P2WPKHTxnBuilder {
...
   }
+
+  /**
+   * Close a vault.
+   *
+   * @param {CloseVault} closeVault txn to create
+   * @param {Script} changeScript to send unspent to after deducting the (converted + fees)
+   * @returns {Promise<TransactionSegWit>}
+   */
+
```
Comment (canonbrother) [inline packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts:185]:
```suggestion
```

File: packages/jellyfish-transaction/src/script/dftx/dftx.ts [hunk 8]
```diff
@@ -91,7 +91,9 @@ import {
   CTakeLoan,
   TakeLoan,
   CPaybackLoan,
-  PaybackLoan
+  PaybackLoan,
+  CloseVault,
+  CCloseVault
```
Comment (canonbrother) [inline packages/jellyfish-transaction/src/script/dftx/dftx.ts:96]: better move to vault section

File: packages/jellyfish-transaction/src/script/dftx/dftx.ts [hunk 9]
```diff
@@ -250,6 +252,8 @@ export class CDfTx extends ComposableBuffer<DfTx<any>> {
         return compose<TakeLoan>(CTakeLoan.OP_NAME, d => new CTakeLoan(d))
       case CPaybackLoan.OP_CODE:
         return compose<PaybackLoan>(CPaybackLoan.OP_NAME, d => new CPaybackLoan(d))
+      case CCloseVault.OP_CODE:
```
Comment (canonbrother) [inline packages/jellyfish-transaction/src/script/dftx/dftx.ts:255]: move to vault section

**Review by canonbrother — COMMENTED** [review]
code organization stuff.. move closeVault to vault section instead after pay loan

**Review by canonbrother — COMMENTED** [review]
overall looks good for me

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts (hunk 7 above)
Comment (surangap) [reply packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts:185]: looks okay? one line break

Comment (canonbrother) [reply packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts:185]: yea.. just extra break line

**Review by surangap — APPROVED** [review]
can do code organization after development. I think we have some rearrangements to do.

**Comment by canonbrother** [comment]
> can do code organization after development. I think we have some rearrangements to do in others as well.

yea no rush for this.. can continue on next pr

**Review by canonbrother — APPROVED** [review]

File: packages/testcontainers/src/containers/RegTestContainer/Masternode.ts [hunk 10]
```diff
@@ -210,6 +224,48 @@ export class MasterNodeRegTestContainer extends RegTestContainer {
...
+
+  /**
+   * Wait for next price
+   *
+   * @param {string} fixedIntervalPriceId
+   * @param {number} nextPrice
+   * @param {number} [timeout=30000] in ms
+   * @return {Promise<void>}
+   */
+  async waitForNextPrice (fixedIntervalPriceId: string, nextPrice: number, timeout = 100000): Promise<void> {
```
Comment (fuxingloh) [inline packages/testcontainers/src/containers/RegTestContainer/Masternode.ts:256]: `activePrice`/`nextPrice` should be parsed `string` as they are precision sensitive

**Review by fuxingloh — APPROVED** [review]

**Comment by fuxingloh** [comment]
going merge first, we can address the requested change later

File: packages/jellyfish-transaction/src/script/dftx/dftx.ts (hunk 8 above)
Comment (jingyi2811) [reply packages/jellyfish-transaction/src/script/dftx/dftx.ts:96]: This is the result of merge conflict. I am going to fix this in another PR.

File: packages/jellyfish-transaction/src/script/dftx/dftx.ts (hunk 9 above)
Comment (jingyi2811) [reply packages/jellyfish-transaction/src/script/dftx/dftx.ts:255]: This is the result of merge conflict. I am going to fix this in another PR.

File: packages/testcontainers/src/containers/RegTestContainer/Masternode.ts (hunk 10 above)
Comment (jingyi2811) [reply packages/testcontainers/src/containers/RegTestContainer/Masternode.ts:256]: Ok. I am going to fix this in another PR.

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts (hunk 7 above)
Comment (jingyi2811) [reply packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts:185]: Ok. I am going to fix this in another PR.

File: packages/jellyfish-transaction/src/script/dftx/dftx.ts (hunk 9 above)
Comment (jingyi2811) [reply packages/jellyfish-transaction/src/script/dftx/dftx.ts:255]: Fixed in updateVault DFTX PR

File: packages/jellyfish-transaction/src/script/dftx/dftx.ts (hunk 8 above)
Comment (jingyi2811) [reply packages/jellyfish-transaction/src/script/dftx/dftx.ts:96]: Fixed in updateVault DFTX PR

File: packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts (hunk 7 above)
Comment (jingyi2811) [reply packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts:185]: Hmm, this seems correct to me. A break line should be added after the comment section.
Anyway, I have group all the functions name ended with `Vault` together

Screenshot 2021-11-09 at 12 18 11 AM

Comment (jingyi2811) [reply packages/jellyfish-transaction-builder/src/txn/txn_builder_loans.ts:185]: Fixed in updateVault DFTX PR

File: packages/testcontainers/src/containers/RegTestContainer/Masternode.ts (hunk 10 above)
Comment (jingyi2811) [reply packages/testcontainers/src/containers/RegTestContainer/Masternode.ts:256]: Fixed in UpdateVault DFTX PR

**Pull request merged**